tests/src/run.rs

```diff
@@ -348,6 +372,47 @@ impl<'a> Runner<'a> {
             "oob".into()
         }
     }
+
+    fn format_pos_external(&self, id: FileId, pos: usize) -> String {
+        // We can unwrap since test has run, these paths are correct
+        let path = system_path(id).unwrap();
+        let bytes = read(&path).unwrap();
+        if pos > bytes.len() {
+            return "oob".into();
+        }
+
+        let text = String::from_utf8_lossy(&bytes);
+
+        let mut line = 1;
```

Comment:
This looks very similar to the Span code. The `Source` type can already do it (and faster than this here!) with a `lines` vector. Maybe we could even turn that into a reusable `Lines` type that is not tied to Typst source files, but can also handle its in-place edits! Especially in the context of the web app, we'll also need the UTF-16 conversion methods to be able to make use of these changes.

We could even consider memoizing the output via `lines(bytes: &Bytes) -> Arc<Lines>` (right now many errors in a long binary file would result in technically O(n^2) runtime).

This is a bit of larger thing and I know you're busy, so don't feel forced to work on this. Maybe I'd pick it up otherwise.